File: readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
	string text = get_string("Text: ");
	int letters = 0, words = 0, sentences = 0;
	float L = 0, S = 0;
	letters = count_letters(text);
	words = count_words(text);
	sentences = count_sentences(text);

	//L is the average number of letters per 100 words in the text.
	L = (float) letters / (float) words * 100;

	//S is the average number of sentences per 100 words in the text.
	S = (float) sentences / (float) words * 100;

	float index = 0;
	index = 0.0588 * L - 0.296 * S - 15.8;

	if (index < 1)
		printf("Before Grade 1\n");
	else if (index >= 16)
		printf("Grade 16+\n");
	else
	{
		printf("Grade %d\n", (int)round(index));
	}


}

int count_letters(string text)
{
	int letters = 0;

	int i = 0;
	char c = '\0';
	while (text[i])
	{
	 	c = text[i];
		//check letters to be uppercase or lowercase alphabetical character
		if (islower(c) || isupper(c))
			letters++;
		i++;
	}
	return letters;
}

int count_words(string text)
{
	int words = 1;

	int i = 0;
	char c = '\0';
	while (text[i])
	{
	 	c = text[i];
		if (c == 32)
		{
			//handle multiple spaces between words
			while (text[i] == 32)
				i++;
			words++;
		}
		i++;
	}
	return words;
}

int count_sentences(string text)
{
	int sentences = 0;

	int i = 0;
	char c = '\0';
	while (text[i])
	{
	 	c = text[i];
		if (c == '.' || c == '!' || c == '?')
			sentences++;
		i++;
	}
	return sentences;
}

File: Week 6 - Python/sentimental-readability/readability.py
from cs50 import get_string
import re


def main():
    # Ask the user to type in some text
    input = get_string("Text: ")

    # Calculate grade according to the Coleman-Liau formula
    grade = calculate_grade(input)

    # Output the grade level for the text
    if grade >= 16:
        print(f"Grade: 16+")
    elif grade < 1:
        print("Before Grade 1")
    else:
        print(f"Grade: {grade}")


def calculate_grade(input):
    nb_letters = count_letters(input)
    nb_words = count_words(input)
    nb_sentences = count_sentences(input)

    # L is the average number of letters per 100 words in the text
    L = (nb_letters * 100) / float(nb_words)

    # S is the average number of sentences per 100 words in the text
    S = (nb_sentences * 100) / float(nb_words)

    # Coleman-Liau index is 0.0588 * L - 0.296 * S - 15.8
    index = 0.0588 * L - 0.296 * S - 15.8

    # Return grade level computed by the Coleman-Liau formula, rounded to the nearest integer
    return round(index)


# Count the number of letters in the text
# A letter is any lowercase character from a to z or any uppercase character from A to Z
def count_letters(text):
    return len(re.findall(r"[a-zA-Z]", text))


# Count the number of words in the text
# Any sequence of characters separated by spaces should count as a word
def count_words(text):
    return len(text.split())


# Count the number of sentences in the text
# Any occurrence of a period, exclamation point, or question mark indicates the end of a sentence
def count_sentences(text):
    return len(re.split(r"[.!?]", text)) - 1


if __name__ == "__main__":
    main()
